src/db/timesheets/ItemsOperations.ts

```diff
@@ -0,0 +1,221 @@
+        }
+    }
+}
+
+
+/*
+    Implementation for processing operations on the TableData (HoursData) of our timesheet 
+    i.e. the user entered rows of the time they worked. 
+*/
+export class HoursDataOperations implements ItemsOperations {
+    public  Insert(timesheet: TimeSheetSchema, body:InsertRequest)  {
+        const data = timesheet.HoursData; 
+
+        const item = TimesheetEntrySchema.parse(body.Item); 
+        // Sorting is currently only day by day based - need some way of minute by minute 
+        var idx = 0; 
+        for (idx; idx < data.length; idx += 1) {
+            const row = data[idx]; 
+            if (moment.unix(row.Date).isAfter(moment.unix(item.Date), 'day')) {
+                break; 
+            }
+        }
+        //Insert into front of list 
+        if (idx === 0) {
+            return {
+                ...timesheet, 
+                HoursData: [
+                    item,
+                    ...data 
+                ]
+            }; 
+        } else if (idx === data.length) {
+            //End of list 
+            return {
+                ...timesheet, 
+                HoursData: [
+                    ...data,
+                    item
+                ]
+            }; 
+        } else {
+            return {
+                ...timesheet, 
+                HoursData: [
+                    ...data.slice(0, idx), 
+                    item, 
+                    ...data.slice(idx + 1 )
+                ]
+            }
+        }
+    }
+    public Delete(timesheet: TimeSheetSchema, body:DeleteRequest)  {
+        return {
+            ...timesheet, 
+            HoursData: timesheet.HoursData.filter((row) => row.EntryID !== body.Id)
+        }; 
+         
+    }
+
+    public  Update(timesheet: TimeSheetSchema, body:UpdateRequest)  {
+        if (timesheet.HoursData?.filter((row) => row.EntryID === body.Id).length === 0) {
+            throw new Error("Could not find a row with that ID"); 
+        }
+        return {
+            ...timesheet, 
+            HoursData: timesheet.HoursData.map((row) => {
+                // Only update the one specific id 
+                if (row.EntryID === body.Id) {
+                    return {
+                        ...row, 
+                        [body.Attribute] : body.Data
+                    }; 
+                }
+                return row; 
+            })
+        }
+    }
+}
+
+// Class for operations on the schedule data field - i.e. the supervisor reported hours they should have worked. 
+export class ScheduledDataOperations implements ItemsOperations {
+    public  Insert(timesheet: TimeSheetSchema, body:InsertRequest)  {
+        const data = timesheet.ScheduleData; 
+        const item = ScheduleEntrySchema.parse(body.Item); 
+        // Sorting is currently only day by day based - need some way of minute by minute 
+        var idx = 0; 
+        for (idx; idx < data.length; idx += 1) {
```

Comment:
Would it be easier to just use the built-in sort functionality, giving it the `moment.unix(row.Date).isAfter(moment.unix(item.Date, 'day'))` as the comparator function? Might be a little less efficient, but could be more readable.

Also, if we define our own comparator function, I think we could sort using both the `moment.isAfter` method and the startDateTime. Thoughts? Doesn't have to be in this PR

Reply:
Or, maybe better, build out a small utility class that has this sorted insert functionality, and call it here, just so we don't have this code block repeated for each of the three operations

Reply:
yeah that makes sense - to make this ticket just merged ill make another ticket for this functionality / extending it to make it based on the minute actual accurate insertion instead of just based on the day.